### PR TITLE
feat: add user registration form validation

### PR DESCRIPTION
**Pull request by jiejasonliu**
## Summary
- sign-up page reorganized form inputs
- sign-up page has form validation
  - only simple validation on client side
  - otherwise, complex ones handled on server (maybe client side should be more strict?)
- sign-in page works
- guard the registration pages if you are already signed in

## Preview

https://user-images.githubusercontent.com/29434693/190529319-ea9eef24-ac94-4d42-8c8a-bc997acedf3c.mp4

### CODE REVIEW

**Comment by derricklee43** [comment]
damn, Fort Knox over here :laughing:

File: src/pages/sign-up/sign-up-page.tsx [hunk 1]
```diff
@@ -2,47 +2,151 @@ import React, { useState } from 'react';
...
+    if (!data) {
+      throw new Error('An unknown error occurred while submitting the registration form');
+    }
+
+    if (data.statusCode === 200) {
+      const { id } = data.response;
+      console.log('Successfully registered:', id);
+      navigate(paths.signIn);
+      // TODO: redirect to sign up success page with button to login
+      // I don't think we should automatically log users in
```
Comment (derricklee43) [inline src/pages/sign-up/sign-up-page.tsx:119]: Is there a reason why we don't let them in immediately after creating an account?

File: src/assets/_colors.scss [hunk 2]
```diff
@@ -14,3 +14,5 @@ $blue: #25283d;
 $blue-hover: #1e1f31;
 $light-placeholder: #636b8b;
 $dark-placeholder: #595c80;
+
+$light-red: #e16060;
```
Comment (derricklee43) [inline src/assets/_colors.scss:18]: That's a good red:joy:

File: src/hooks/api/use-user-client.ts [hunk 3]
```diff
@@ -20,11 +27,51 @@ export function useUserClient() {
...
+      const payload = registerUserInfoToJson(userInfo);
+      const response = await fetchWrapper.post('/Register', payload, numRetries);
+      if (isRegisterSuccess(response)) {
+        return { statusCode: 200, response };
+      } else {
+        throw new Error('Expected register success to have id (userId), but got none');
+      }
+    } catch (error) {
+      // data is an array of IdentityErrors
+      debugger;
```
Comment (derricklee43) [inline src/hooks/api/use-user-client.ts:64]: Did you mean to keep this in?

File: src/pages/sign-in-page/sign-in-page.tsx [hunk 4]
```diff
@@ -29,7 +31,17 @@ export const SignInPage = () => {
   );
 
   function handleSubmitClick() {
-    console.log(userName, password);
+    userClient.login(userName, password).then((success) => {
```
Comment (derricklee43) [inline src/pages/sign-in-page/sign-in-page.tsx:34]: Is there a reason here for using `.then` instead of `await`?

**Review by derricklee43 — APPROVED** [review]
I like it! LGMT!

File: src/pages/sign-up/sign-up-page.tsx (hunk 1 above)
Comment (jiejasonliu) [reply src/pages/sign-up/sign-up-page.tsx:119]: i think this applies more to apps that have you verify an email + ability to change password

it's usually a good idea to let users try to login but it's different here i guess
no preference really so i'll change it to just log them in

File: src/hooks/api/use-user-client.ts (hunk 3 above)
Comment (jiejasonliu) [reply src/hooks/api/use-user-client.ts:64]: of course :trollface:

File: src/pages/sign-up/sign-up-page.tsx (hunk 1 above)
Comment (derricklee43) [reply src/pages/sign-up/sign-up-page.tsx:119]: oh, I see. I'm down with keeping it as is. Either way

File: src/pages/sign-in-page/sign-in-page.tsx (hunk 4 above)
Comment (jiejasonliu) [reply src/pages/sign-in-page/sign-in-page.tsx:34]: no, making the function async and awaiting this is equivalent
but i'll change it to be consistent w our other async handlers

File: src/pages/sign-up/sign-up-page.tsx [hunk 5]
```diff
@@ -2,47 +2,151 @@ import React, { useState } from 'react';
...
+          newFormError.email = desc;
+        } else if (code.includes('Password')) {
+          newFormError.password = desc;
+        }
+      });
+      setFormError(newFormError);
+    }
+  }
+
+  function _getLabelError(
```
[the remark this answers is not in the document]
Comment (derricklee43) [reply src/pages/sign-up/sign-up-page.tsx]: oh, I see. I like that. I think I'll start doing that for helpers

**Comment by derricklee43** [comment]
lgtm

**Pull request merged**